jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/params/DefaultXGBoostParamsWriter.scala

```diff
@@ -0,0 +1,215 @@
+    *  - paramMap
+    *  - (optionally, extra metadata)
+    *
+    * @param extraMetadata Extra metadata to be saved at same level as uid, paramMap, etc.
+    * @param paramMap      If given, this is saved in the "paramMap" field.
+    *                      Otherwise, all [[org.apache.spark.ml.param.Param]]s are encoded using
+    *                      [[org.apache.spark.ml.param.Param.jsonEncode()]].
+    */
+  def saveMetadata(
+                    instance: Params,
+                    path: String,
+                    sc: SparkContext,
+                    extraMetadata: Option[JObject] = None,
+                    paramMap: Option[JValue] = None): Unit = {
+    val metadataPath = new Path(path, "metadata").toString
+    val metadataJson = getMetadataToSave(instance, sc, extraMetadata, paramMap)
+    sc.parallelize(Seq(metadataJson), 1).saveAsTextFile(metadataPath)
+  }
+
+  /**
+    * Helper for [[saveMetadata()]] which extracts the JSON to save.
+    * This is useful for ensemble models which need to save metadata for many sub-models.
+    *
+    * @see [[saveMetadata()]] for details on what this includes.
+    */
+  def getMetadataToSave(
+                         instance: Params,
+                         sc: SparkContext,
+                         extraMetadata: Option[JObject] = None,
+                         paramMap: Option[JValue] = None): String = {
+    val uid = instance.uid
+    val cls = instance.getClass.getName
+    val params = instance.extractParamMap().toSeq.asInstanceOf[Seq[ParamPair[Any]]]
+    val jsonParams = paramMap.getOrElse(render(params.map { case ParamPair(p, v) =>
+      p.name -> parse(p.jsonEncode(v))
+    }.toList))
+    val basicMetadata = ("class" -> cls) ~
+      ("timestamp" -> System.currentTimeMillis()) ~
+      ("sparkVersion" -> sc.version) ~
+      ("uid" -> uid) ~
+      ("paramMap" -> jsonParams)
+    val metadata = extraMetadata match {
+      case Some(jObject) =>
+        basicMetadata ~ jObject
+      case None =>
+        basicMetadata
+    }
+    val metadataJson: String = compact(render(metadata))
+    metadataJson
+  }
+}
+
+/**
+  * Default [[MLReader]] implementation for transformers and estimators that contain basic
+  * (json4s-serializable) params and no data. This will not handle more complex params or types with
+  * data (e.g., models with coefficients).
+  *
+  * @tparam T ML instance type
+  *           TODO: Consider adding check for correct class name.
+  */
+private[xgboost4j] class DefaultXGBoostParamsReader[T] extends MLReader[T] {
+
+  override def load(path: String): T = {
+    val metadata = DefaultXGBoostParamsReader.loadMetadata(path, sc)
+    val cls = classForName(metadata.className)
+    val instance =
+      cls.getConstructor(classOf[String]).newInstance(metadata.uid).asInstanceOf[Params]
+    DefaultXGBoostParamsReader.getAndSetParams(instance, metadata)
+    instance.asInstanceOf[T]
+  }
+}
+
+private[xgboost4j] object DefaultXGBoostParamsReader {
```

Comment:
Do you need this though?
The `Pipeline` object will call your customized reader, once the metadata is loaded. 

I guess you only need the DefaultParamsWriter, but I could be wrong.